topology/nhlt/intel/dmic/dmic-process.c

```diff
@@ -0,0 +1,1341 @@
+	int di;
+
+	if (!dmic)
+		return -EINVAL;
+
+	if (pdm_index >= DMIC_HW_CONTROLLERS) {
+		fprintf(stderr, "set_pdm_data illegal pdm_index\n");
+		return -EINVAL;
+	}
+
+	di = dmic->dmic_dai_index;
+
+	dmic->dmic_prm[di].pdm[pdm_index].enable_mic_a = enable_a;
+	dmic->dmic_prm[di].pdm[pdm_index].enable_mic_b = enable_b;
+	dmic->dmic_prm[di].pdm[pdm_index].polarity_mic_a = polarity_a;
+	dmic->dmic_prm[di].pdm[pdm_index].polarity_mic_b = polarity_b;
+	dmic->dmic_prm[di].pdm[pdm_index].clk_edge = clk_edge;
+	dmic->dmic_prm[di].pdm[pdm_index].skew = skew;
+
+	dmic->pdm_enabled[pdm_index] = 1;
+
+	return 0;
+}
+
+int dmic_set_ext_params(struct intel_nhlt_params *nhlt, uint32_t snr, uint32_t sensitivity)
+{
+	struct intel_dmic_params *dmic = (struct intel_dmic_params *)nhlt->dmic_params;
+
+	if (!dmic)
+		return -EINVAL;
+
+	dmic->dmic_mic_config.extension = 1;
+	dmic->dmic_mic_config.snr = snr;
+	dmic->dmic_mic_config.sensitivity = sensitivity;
+
+	return 0;
+}
+
+int dmic_set_mic_params(struct intel_nhlt_params *nhlt, int index,
+			uint8_t type, uint8_t panel, uint32_t speaker_position_distance,
+			uint32_t horizontal_offset, uint32_t vertical_offset,
+			uint8_t frequency_low_band, uint8_t frequency_high_band,
+			uint16_t direction_angle, uint16_t elevation_angle,
+			uint16_t vertical_angle_begin, uint16_t vertical_angle_end,
+			uint16_t horizontal_angle_begin, uint16_t horizontal_angle_end)
+{
+	struct intel_dmic_params *dmic = (struct intel_dmic_params *)nhlt->dmic_params;
+
+	if (!dmic)
+		return -EINVAL;
+
+	dmic->dmic_mic_config.vendor[index].type = type;
+	dmic->dmic_mic_config.vendor[index].panel = panel;
+	dmic->dmic_mic_config.vendor[index].speaker_position_distance = speaker_position_distance;
+	dmic->dmic_mic_config.vendor[index].horizontal_offset = horizontal_offset;
+	dmic->dmic_mic_config.vendor[index].vertical_offset = vertical_offset;
+	dmic->dmic_mic_config.vendor[index].frequency_low_band = frequency_low_band;
+	dmic->dmic_mic_config.vendor[index].frequency_high_band = frequency_high_band;
+	dmic->dmic_mic_config.vendor[index].direction_angle = direction_angle;
+	dmic->dmic_mic_config.vendor[index].elevation_angle = elevation_angle;
+	dmic->dmic_mic_config.vendor[index].vertical_angle_begin = vertical_angle_begin;
+	dmic->dmic_mic_config.vendor[index].vertical_angle_end = vertical_angle_end;
+	dmic->dmic_mic_config.vendor[index].horizontal_angle_begin = horizontal_angle_begin;
+	dmic->dmic_mic_config.vendor[index].horizontal_angle_end = horizontal_angle_end;
+
+	dmic->dmic_mic_config.num_mics++;
+
+	return 0;
+}
+
+/* init dmic parameters, should be called before parsing dais */
+int dmic_init_params(struct intel_nhlt_params *nhlt)
+{
+	struct intel_dmic_params *dmic;
+	int i;
+
+	dmic = calloc(1, sizeof(struct intel_dmic_params));
+	if (!dmic)
+		return -ENOMEM;
+
+	nhlt->dmic_params = dmic;
+	dmic->dmic_blob.gateway_attributes = 1;
+	dmic->dmic_blob.clock_on_delay = 16;
```

Comment:
magic number?

Reply:
yeah also an item we dont have in topology. According to @singalsu could be even 50ms. So this might be handled differently in different fw variants, as for .e.g in sof we have volume ramp to cover for dmic settling down. But will mark this with some define or comments.